### ISSUE TITLE
Change Password issue: AD password change failed, but keychain password was changed

### ISSUE THREAD

**Issue by steven-j-klein** [macmule/ADPassMon#41]
My Preferences are set to Allow password changes, using ADPassMon."

A few minutes ago I used ADPassMon to change my password from **Password1** to **Password2**. Since our AD policy restricts password changes to once every 24 hours, I decided to try to change it again to see how ADPassMon would handle it when our DC rejected the change.

In the _Old Password_ field I entered **Password2**, and in _New Password_ and _Verify Password_ I entered **Password3** and clicked **Change**. As expected, it failed. I saw this error: **Password change failed. Please try again.**

I clicked **OK** and then saw another dialog saying the change was **successful**.

Huh?

I closed the _Change Password_ dialog, then tried to use _Change Password…_ a 3ʳᵈ time, and got a message that my keychain was locked.

Huh?

Using _Keychain Access_ I verified that it was indeed locked. And when I tried to manually unlock it, I discovered that I had to use **Password3**.

It took me a minute to figure out what happened. Even though the AD password change failed (my AD password remained **Password2**), ADPassMon nevertheless changed my login keychain password to **Password3**!

In other words, ADPassMon had created a situation in which **my AD password was different from my login keychain password.**

Avoiding exactly that problem is the main reason I want deploy ADPassMon.

**Comment by macmule** [comment]
Thank for this.

I'll have a nose & will see if I can push a fix next week.

**Comment by steven-j-klein** [comment]
Thank _you_ for creating a terrific tool! (And for being so quick to respond to error reports like this.)

**Comment by macmule** [comment]
Change the below from line 1082:

```
    -- Attempt change password -  the meat of v2 behaviour
    on attemptChangePassword_(sender)
        -- If changing password, change keychain pass too.
        if passwordPromptWindowButton1 is equal to "Change"
            updatePassword_(me)
            updateKeychainPassword_(me)
        else
            set userPasswordChanged to true
            updateKeychainPassword_(me)
        end if
    end attemptChangePassword_

    -- Try & reset the users password via dscl
    on updatePassword_(sender)
        try
            log "Attempting user password change.."
            do shell script "dscl . -passwd /Users/$USER " & quoted form of enteredOldPassword & " " & quoted form of enteredNewPassword
            log "Password changed!"
            set userPasswordChanged to true
            -- Set Keychain settings to make sure they are unlocked
            setKeychainSettings_(me)
```

to:

```
    -- Attempt change password -  the meat of v2 behaviour
    on attemptChangePassword_(sender)
        -- If changing password, change keychain pass too.
        if passwordPromptWindowButton1 is equal to "Change"
            updatePassword_(me)
        else
            set userPasswordChanged to true
            updateKeychainPassword_(me)
        end if
    end attemptChangePassword_

    -- Try & reset the users password via dscl
    on updatePassword_(sender)
        try
            log "Attempting user password change.."
            do shell script "dscl . -passwd /Users/$USER " & quoted form of enteredOldPassword & " " & quoted form of enteredNewPassword
            log "Password changed!"
            set userPasswordChanged to true
            -- Set Keychain settings to make sure they are unlocked
            setKeychainSettings_(me)
            updateKeychainPassword_(me)
```

Simply moving updateKeychainPassword_(me) till only being ran when a successful password change is made.

**Comment by macmule** [comment]
@steven-j-klein Please test https://github.com/macmule/ADPassMon/releases/tag/2.20.17

**Comment by macmule** [comment]
2.20.18 released with the above solution in it.
